### PR TITLE
[PDI-13961] - Spoon firstly tries to conect via JNDI although connection type is JDBC

### PR DESCRIPTION
**Pull request by akhayrutdinov**
- introduce a new method for distinguishing the source where to look for data sources
- remove some legacy code, that is not used anymore
- add tests

@mattyb149, @brosander, @mattcasters, @deinspanjer, @pentaho-nbaker, @rmansoor,

Please review it and mention your opinions. This is a risky change, as it touches the core class.

/cc @pamval

### CODE REVIEW

File: core/src/org/pentaho/di/core/database/DataSourceProviderInterface.java [hunk 1]
```diff
@@ -38,6 +38,19 @@
    * @param datasourceName
    * @return javax.sql.DataSource
    */
-  public DataSource getNamedDataSource( String datasourceName ) throws DataSourceNamingException;
```
Comment (deinspanjer) [inline core/src/org/pentaho/di/core/database/DataSourceProviderInterface.java]: Why the change to package private here?

Comment (akhayrutdinov) [inline core/src/org/pentaho/di/core/database/DataSourceProviderInterface.java]: @deinspanjer, it is redundant, as ```DataSourceProviderInterface``` is interface :)

Comment (deinspanjer) [inline core/src/org/pentaho/di/core/database/DataSourceProviderInterface.java]: Sorry, I overlooked the fact this was a method, thought it was changing the interface itself to package private.

**Comment by deinspanjer** [comment]
A simple code review didn't raise any red flags for me.  I'll need to do some actual testing on Monday.

File: core/src/org/pentaho/di/core/database/DataSourceProviderInterface.java [hunk 2]
```diff
@@ -38,6 +38,19 @@
...
 
+  /**
+   * Returns the named data source of respecting its <code>type</code>
+   *
+   * @param datasourceName name of the desired data source
+   * @param type           data source's type
+   * @return named data source
+   * @throws DataSourceNamingException
+   */
+  DataSource getNamedDataSource( String datasourceName, DatasourceType type ) throws DataSourceNamingException;
```
Comment (mattyb149) [inline core/src/org/pentaho/di/core/database/DataSourceProviderInterface.java]: Quite a few projects use this interface, it is causing downstream failures

**Comment by brendan-lynch** [comment]
This change will cause classes downstream to break such as org.pentaho.platform.plugin.action.kettle.PlatformKettleDataSourceProvider who implement the interface.

**Comment by deinspanjer** [comment]
The PR is supposed to be paired with pentaho/pentaho-platform#2453 . I notified @rmansoor and @pentaho-nbaker based on an e-mail Friday about this, but it looks like I should have just +1'ed it and waited for the other half to be merged. :/  I've backed it out for now in 
https://github.com/pentaho/pentaho-kettle/issues/1660
